File: frontend/src/types/policy.ts
// frontend/src/types/policy.ts — TypeScript Interfaces for HR Policy Assistant

export interface EmployeeRecord {
  employee_id: string;
  name: string;
  job_title: string;
  department: string;
  duty_station: string;
  jurisdiction: "Kenya" | "Ireland" | "Cote d'Ivoire" | "Rwanda" | "Global";
  tenure_months: number;
  employment_status: "Confirmed" | "Probation";
  annual_leave_balance: number;
  basic_salary_monthly: number;
  separation_reason?: string | null;
}

export interface PolicyQueryRequest {
  employee_id: string;
  question: string;
  case_id?: string;
  top_k?: number;
  temperature?: number;
  model?: string;
}

export interface OllamaModelListResponse {
  default_model: string;
  models: string[];
  agent_models: string[];
}

export interface ToolCallRecord {
  step: number;
  tool_name: string;
  arguments: Record<string, any>;
  output: any;
  latency_ms: number;
}

export interface RetryRecord {
  attempt: number;
  status: "SUCCESS" | "RETRY" | "FAILED" | "BUDGET_EXHAUSTED";
  retry_reason?: string | null;
  retryable: boolean;
  latency_ms: number;
  input_tokens: number;
  output_tokens: number;
  total_tokens: number;
  estimated_cost: number;
  is_retry: boolean;
}

export interface RoutingDecision {
  mode: "workflow" | "agent";
  complexity: "SIMPLE" | "MODERATE" | "COMPLEX";
  reason: string;
  requires_agent: boolean;
  routing_ms: number;
  routing_id: string;
  matched_signals: string[];
}

export interface PolicyOutputContract {
  // Identification
  case_id: string;
  employee_id: string;
  question: string;
  run_id?: string;
  evaluation_type?: string;

  // Result
  entitlement_value: string;
  rule_cited: string;
  explanation: string;
  passed: boolean;
  implementation: "agent" | "workflow";

  // Routing metadata
  execution_mode?: "workflow" | "agent";
  routing_reason?: string;
  complexity?: "SIMPLE" | "MODERATE" | "COMPLEX";
  routing_ms?: number;
  mode_history?: string[];
  routing?: RoutingDecision;

  // Execution telemetry
  tool_calls: ToolCallRecord[];
  iterations: number;

  // Token accounting
  prompt_tokens: number;
  completion_tokens: number;
  total_tokens: number;
  token_source?: string; // "ollama_live" | "proxy_estimate" | "unavailable"
  llm_calls?: any[];

  // Cost
  cost_usd: number;
  provider_cost?: string; // "N/A" for local Ollama

  // Latency
  latency_ms: number;
  router_latency_ms?: number;
  execution_latency_ms?: number;

  // Retry metadata
  attempt?: number;
  max_retries?: number;
  total_attempts?: number;
  retry_history?: RetryRecord[];

  // Termination
  termination_reason: string;

  // Config provenance
  top_k?: number | null;
  temperature?: number | null;
  model?: string | null;
}

export interface BenchmarkCase {
  case_id: string;
  employee_id: string;
  question: string;
  source_section: string;
  expected_value: string;
  tenure_dependency?: string;
  deterministic_pass_criteria: string[];
}

export interface ExecutionSummary {
  pass_rate_pct: number;
  passed_count: number;
  total_cases: number;
  p50_latency_ms: number;
  total_tokens: number;
  cost_per_question_usd: number;
}

export interface PolicyBenchmarkResponse {
  summary: {
    agent: ExecutionSummary;
    workflow: ExecutionSummary;
  };
  agent_results: PolicyOutputContract[];
  workflow_results: PolicyOutputContract[];
}

export interface BenchmarkCaseLiveStatus {
  case_id: string;
  employee_id: string;
  question: string;
  ground_truth?: string;
  source_section?: string;
  pass_criteria?: string[];
  status: "WAITING" | "RUNNING" | "PASS" | "FAIL" | "ERROR";
  agent_status: "WAITING" | "RUNNING" | "PASS" | "FAIL" | "ERROR";
  workflow_status: "WAITING" | "RUNNING" | "PASS" | "FAIL" | "ERROR";
  agent_entitlement?: string | null;
  workflow_entitlement?: string | null;
  agent_rule?: string | null;
  workflow_rule?: string | null;
  agent_explanation?: string | null;
  workflow_explanation?: string | null;
  agent_passed?: boolean | null;
  workflow_passed?: boolean | null;
  agent_latency_ms?: number | null;
  workflow_latency_ms?: number | null;
  agent_tokens?: number | null;
  workflow_tokens?: number | null;
  agent_cost_usd?: number | null;
  workflow_cost_usd?: number | null;
  agent_result?: PolicyOutputContract | null;
  workflow_result?: PolicyOutputContract | null;
}

export interface PolicyBenchmarkRunStateResponse {
  run_id: string;
  status: "RUNNING" | "COMPLETED" | "CANCELLED" | "ERROR";
  top_k: number;
  temperature: number;
  model: string;
  total_cases: number;
  completed_cases: number;
  progress_pct: number;
  current_case_id?: string | null;
  current_question?: string | null;
  current_agent_stage?: string | null;
  current_workflow_stage?: string | null;
  agent_completed_count: number;
  workflow_completed_count: number;
  elapsed_seconds: number;
  cases_status: BenchmarkCaseLiveStatus[];
  summary?: {
    agent: ExecutionSummary;
    workflow: ExecutionSummary;
  } | null;
  error_message?: string | null;
  agent_results?: PolicyOutputContract[];
  workflow_results?: PolicyOutputContract[];
}

export interface ActiveBenchmarkRunResponse {
  active: boolean;
  run: PolicyBenchmarkRunStateResponse | null;
}

File: backend/schemas/policy.py
# backend/schemas/policy.py — Typed Pydantic Contracts and Schemas for HR Policy Execution
from __future__ import annotations

from enum import Enum
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field

# Strict Execution Budgets & Constants
MAX_ITERATIONS: int = 5
MAX_TOKENS: int = 4000
MAX_COST: float = 0.05
MAX_WALL_CLOCK_SECONDS: float = 45.0
MAX_RETRIES: int = 2                     # Maximum automatic retries (initial attempt + 2 = 3 total)

# Token cost proxy — single source of truth, never scattered through code.
# Local Ollama has no direct provider billing cost.
# This proxy rate ($0.50/1M tokens) is used only for educational cost comparison.
# UI must label this "Estimated Token Cost", not "Actual Cost".
TOKEN_COST_PER_1M: float = 0.50          # dollars per 1 million tokens
TOKEN_COST_PROXY_RATE: float = TOKEN_COST_PER_1M / 1_000_000  # = 0.000000_5


class JurisdictionEnum(str, Enum):
    KENYA = "Kenya"
    IRELAND = "Ireland"
    COTE_D_IVOIRE = "Cote d'Ivoire"
    RWANDA = "Rwanda"
    GLOBAL = "Global"


class PolicyCategoryEnum(str, Enum):
    LEAVE = "leave"
    NOTICE_AND_SEPARATION = "notice_and_separation"
    BENEFITS_AND_PENSION = "benefits_and_pension"
    HOLIDAYS_AND_WORKING_HOURS = "holidays_and_working_hours"
    CONDUCT_AND_DISCIPLINE = "conduct_and_discipline"


class EmployeeRecord(BaseModel):
    employee_id: str
    name: str
    job_title: str
    department: str
    duty_station: str
    jurisdiction: JurisdictionEnum
    tenure_months: int
    employment_status: str
    annual_leave_balance: int = 0
    basic_salary_monthly: float = 0.0
    separation_reason: Optional[str] = None


class PolicyQueryRequest(BaseModel):
    employee_id: str = Field(..., description="Unique Employee ID, e.g. EMP001")
    question: str = Field(..., description="The policy entitlement or rule query")
    case_id: Optional[str] = Field(None, description="Optional benchmark case ID")
    top_k: Optional[int] = Field(5, description="Top K retrieval count")
    temperature: Optional[float] = Field(0.3, description="LLM sampling temperature")
    model: Optional[str] = Field("llama3.1:8b", description="Model name")


# Telemetry record for a single LLM call (including retries)
class LLMCallRecord(BaseModel):
    call_index: int                  # 1-indexed within the run (retries count separately)
    attempt: int                     # 1 = initial, 2 = retry 1, 3 = retry 2
    is_retry: bool = False
    input_tokens: int = 0
    output_tokens: int = 0
    total_tokens: int = 0
    latency_ms: float = 0.0
    token_source: str = "ollama_live"  # "ollama_live" | "proxy_estimate" | "unavailable"
    status: str = "SUCCESS"           # "SUCCESS" | "RETRY" | "FAILED"
    retry_reason: Optional[str] = None
    retryable: bool = True


# Retry attempt record
class RetryRecord(BaseModel):
    attempt: int                     # Which attempt number (1=initial, 2=retry1, 3=retry2)
    status: str                      # "SUCCESS" | "RETRY" | "FAILED" | "BUDGET_EXHAUSTED"
    retry_reason: Optional[str] = None
    retryable: bool = True
    latency_ms: float = 0.0
    input_tokens: int = 0
    output_tokens: int = 0
    total_tokens: int = 0
    estimated_cost: float = 0.0


class PolicyOutputContract(BaseModel):
    # Identification
    case_id: str
    employee_id: str
    question: str
    run_id: str = ""
    evaluation_type: str = "WEEK7_POLICY_EXECUTION"

    # Result
    entitlement_value: str
    rule_cited: str
    explanation: str
    passed: bool = False
    implementation: str = "agent"      # "agent" | "workflow"

    # Routing metadata
    execution_mode: str = "workflow"   # "workflow" | "agent"
    routing_reason: str = ""
    complexity: str = "SIMPLE"
    routing_ms: float = 0.0
    mode_history: List[str] = Field(default_factory=list)  # tracks mode switches

    # Execution telemetry
    tool_calls: List[Dict[str, Any]] = Field(default_factory=list)
    iterations: int = 1

    # Token accounting — real Ollama metadata preferred
    prompt_tokens: int = 0
    completion_tokens: int = 0
    total_tokens: int = 0
    token_source: str = "unavailable"  # "ollama_live" | "proxy_estimate" | "unavailable"
    llm_calls: List[Dict[str, Any]] = Field(default_factory=list)  # per-call breakdown

    # Cost — labeled as estimated, never as actual billing
    cost_usd: float = 0.0              # estimated token cost
    provider_cost: str = "N/A"         # always "N/A" for local Ollama

    # Latency
    latency_ms: float = 0.0            # total request latency via time.perf_counter()
    router_latency_ms: float = 0.0
    execution_latency_ms: float = 0.0

    # Retry metadata
    attempt: int = 1                   # which attempt succeeded (1 = no retry needed)
    max_retries: int = MAX_RETRIES
    total_attempts: int = 1
    retry_history: List[Dict[str, Any]] = Field(default_factory=list)

    # Termination
    termination_reason: str = "SUCCESS"

    # Config provenance
    top_k: Optional[int] = None
    temperature: Optional[float] = None
    model: Optional[str] = None


class BenchmarkCase(BaseModel):
    case_id: str
    employee_id: str
    question: str
    source_section: str
    expected_value: str
    tenure_dependency: Optional[str] = None
    deterministic_pass_criteria: List[str] = Field(default_factory=list)
